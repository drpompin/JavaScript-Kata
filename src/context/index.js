import React, { createContext, useState } from 'react';

export const KataContext = createContext(null);



export default function Context({ children }) {
    const [kataState, setKataState] = useState([]);


    const handleAddToCart = (item) => {
        let ItemExistsInCart = kataState.length && kataState.find(product => product.name === item.name);
       
        if (ItemExistsInCart) {
            ItemExistsInCart.qty += 1;
            if (!item.specialPrice) {
                ItemExistsInCart.total = ItemExistsInCart.qty * ItemExistsInCart.unitPrice;
            } else {
                if (ItemExistsInCart.qty < +(ItemExistsInCart.specialPrice.split(" ")[0])) {
                    ItemExistsInCart.total = ItemExistsInCart.qty * ItemExistsInCart.unitPrice;
                }

                if (ItemExistsInCart.qty >= +(ItemExistsInCart.specialPrice.split(" ")[0])) {
                    let DiscountableQty = Math.trunc(ItemExistsInCart.qty / +(ItemExistsInCart.specialPrice.split(" ")[0]));
                    let discountedAmount = DiscountableQty * +(ItemExistsInCart.specialPrice.split(" ")[2]);

                    let remainderQty = ItemExistsInCart.qty % +(ItemExistsInCart.specialPrice.split(" ")[0]);
                    let remainderAmount = remainderQty * ItemExistsInCart.unitPrice;
                    
                    ItemExistsInCart.total = (discountedAmount + remainderAmount);

                }
            }
            setKataState ([
                ...kataState
            ]);

        } else {
            setKataState([
                ...kataState,
                {
                    ...item,
                    qty: 1,
                    total: item.unitPrice
                }
            ])
        }
    }

    return (
        <KataContext.Provider value={{ 
            kataState, 
            setKataState, 
            handleAddToCart 
        }}>
            {children}
        </KataContext.Provider>
    )
}
