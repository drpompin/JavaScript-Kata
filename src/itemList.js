import HawaiPizza from './images/hawaiian-pizza.jpg'
import SeaFoodPizza from './images/seafood-pizza.jpg'

export const itemList = [
    {
        name: 'Item A',
        unitPrice: 50,
        specialPrice: '3 for 130',
        image: HawaiPizza
    },
    {
        name: 'Item B',
        unitPrice: 30,
        specialPrice: '2 for 45',
        image: SeaFoodPizza
    },
    {
        name: 'Item C',
        unitPrice: 20,
        specialPrice: '',
        image: HawaiPizza
    },
    {
        name: 'Item D',
        unitPrice: 15,
        specialPrice: '',
        image: SeaFoodPizza
    }
]