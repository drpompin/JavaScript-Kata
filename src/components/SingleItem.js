import React, { useContext, useState, useEffect } from 'react'
import Image from 'next/image'
import { CurrencyFormatter } from '../utilities/currencyFormatter';
import ButtonComponent from './Button'
import { KataContext } from '../../context';



export default function SingleItem(props) {
    const { name, image, unitPrice, specialPrice } = props.item;
	const [selectedItemId, setSelectedItemId] = useState('');
	const [itemIsAdded, setItemIsAdded] = useState(false);

    const { handleAddToCart } = useContext(KataContext);

    const initiateAddToCart = (name, unitPrice, specialPrice, image) => {
        handleAddToCart({
            name,
            unitPrice,
            specialPrice, 
            image
        });
        setSelectedItemId(name);
        setItemIsAdded(true);
    };

    useEffect(() => {
        itemIsAdded &&
        setTimeout(() => {
            setItemIsAdded(false);
            setSelectedItemId("");
        }, 1000);
    }, [itemIsAdded])

    return (
        <div 
            key={name} 
            className="rounded-lg border border-[#ccc] bg-white overflow-hidden py-8 px-4 flex flex-col justify-center items-center cursor-pointer hover:shadow-lg relative"
            data-testid="singleItem"
        >
            <Image src={image} alt='item-image' className='w-full' />
            <div className='text-[#27abce] mt-2 font-semibold'>{name}</div>
            <div className='font-semibold text-sm my-2'>
                {CurrencyFormatter(unitPrice)} each
            </div>

            {   specialPrice ? 
                <div className='font-semibold text-sm mb-2'>{CurrencyFormatter(Number(specialPrice?.split(" ")[2]))} for {specialPrice?.split(" ")[0]}</div>
                :
                <div className='h-5 mb-2'></div>
            }

            <ButtonComponent 
                title="Add To Cart"
                onClick={() => initiateAddToCart(name, unitPrice, specialPrice, image)}
                dataTestid="addItem"
            />

            {itemIsAdded && name === selectedItemId && (
                <div className='absolute left-0 text-[#f52421] top-[90%] w-full text-center'>Item added to cart</div>
            )}
        </div>
    )
}
