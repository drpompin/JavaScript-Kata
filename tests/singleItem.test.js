import React from "react";
import "@testing-library/jest-dom";
import { render } from "@testing-library/react";
import SingleItem from "../src/components/SingleItem";
import { itemList } from "../src/itemList";
const item = itemList[0]



describe('Single Item Component', () => {
test('renders single item correctly', () => {
    const { asFragment } = render(
      <SingleItem item={item} />
    );
    expect(asFragment()).toMatchSnapshot();
  });
})
