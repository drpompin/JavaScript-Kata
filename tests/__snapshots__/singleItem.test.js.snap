// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Single Item Component renders single item correctly 1`] = `
<DocumentFragment>
  <div
    class="rounded-lg border border-[#ccc] bg-white overflow-hidden py-8 px-4 flex flex-col justify-center items-center cursor-pointer hover:shadow-lg relative"
    data-testid="singleItem"
  >
    <img
      alt="item-image"
      class="w-full"
      data-nimg="1"
      decoding="async"
      height="40"
      loading="lazy"
      src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
      srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
      style="color: transparent;"
      width="40"
    />
    <div
      class="text-[#27abce] mt-2 font-semibold"
    >
      Item A
    </div>
    <div
      class="font-semibold text-sm my-2"
    >
      50 Pence each
    </div>
    <div
      class="font-semibold text-sm mb-2"
    >
      Â£1.3 for 3
    </div>
    <button
      class="w-full py-2 font-semibold border-none rounded-md text-sm bg-[#f52421] hover:bg-[#ef3b39] text-white undefined "
      data-testid="addItem"
    >
      Add To Cart
    </button>
  </div>
</DocumentFragment>
`;
